### PR TITLE
Forbid some empty separated lists in v2

### PR DESCRIPTION
**Pull request by Xanewok**
Part of #652

These are a bug-fixes along with regression tests added for each case, we use a definition from v0/v1 for these.

### CODE REVIEW

File: crates/solidity/inputs/language/src/definition.rs [hunk 1]
```diff
@@ -2417,12 +2415,18 @@ codegen_language_macros::compile!(Language(
                         ),
                         Struct(
                             name = OverrideSpecifier,
+                            fields = (
+                                override_keyword = Required(Terminal([OverrideKeyword])),
+                                overriden = Optional(kind = NonTerminal(OverridePathsDeclaration))
```
Comment (OmarTawfik) [inline crates/solidity/inputs/language/src/definition.rs:2420]: nit: spelling
```suggestion
                                overridden = Optional(kind = NonTerminal(OverridePathsDeclaration))
```

File: crates/solidity/inputs/language/src/definition.rs [hunk 2]
```diff
@@ -2417,12 +2415,18 @@ codegen_language_macros::compile!(Language(
...
+                        Struct(
+                            name = OverridePathsDeclaration,
                             error_recovery = FieldsErrorRecovery(
                                 delimiters =
                                     FieldDelimiters(open = open_paren, close = close_paren)
                             ),
                             fields = (
-                                override_keyword = Required(Terminal([OverrideKeyword])),
                                 open_paren = Required(Terminal([OpenParen])),
                                 overridden = Required(NonTerminal(OverridePaths)),
```
Comment (OmarTawfik) [inline crates/solidity/inputs/language/src/definition.rs:2431]: nit: I suggest avoiding `specifier.overridden.unwrap().overridden` if possible. Maybe changing the second one to `paths` instead? `specifier.overridden.unwrap().paths`
```suggestion
                                paths = Required(NonTerminal(OverridePaths)),
```

File: crates/solidity/testing/snapshots/cst_output/TupleDeconstructionStatement/empty/generated/0.4.11-success.yml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# This file is generated automatically by infrastructure scripts. Please don't edit by hand.
+
+Source: >
+  1  │ // TODO: Tuple expressions can't be empty                                        │ 0..41
+  2  │ () = ();                                                                         │ 42..50
```
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/TupleDeconstructionStatement/empty/generated/0.4.11-success.yml:5]: This should have failed. Did I miss something?

File: crates/solidity/testing/snapshots/cst_output/UsingDirective/destructure_empty/input.sol [hunk 4]
```diff
@@ -0,0 +1 @@
+using {} for *;
```
Comment (OmarTawfik) [inline crates/solidity/testing/snapshots/cst_output/UsingDirective/destructure_empty/input.sol:1]: nit: suggest adding the helpful comment like the others:

```suggestion
// empty symbols list is not valid.
using {} for *;
```

**Review by OmarTawfik — CHANGES_REQUESTED** [review]
Left a couple of questions

File: crates/solidity/testing/snapshots/cst_output/TupleDeconstructionStatement/empty/generated/0.4.11-success.yml (hunk 3 above)
Comment (Xanewok) [reply crates/solidity/testing/snapshots/cst_output/TupleDeconstructionStatement/empty/generated/0.4.11-success.yml:5]: This is parsed incorrectly on `main` atm, as we can't express "separated list of optional items, but having at least 1 item" in terms of the "separated" grammar node, short of `(elem (separator elem )*)`.

We should fix it and I'll open an issue for it; wanted to leave a TODO note explaining that the current test case success is not actually expected.

File: crates/solidity/testing/snapshots/cst_output/UsingDirective/destructure_empty/input.sol (hunk 4 above)
Comment (Xanewok) [reply crates/solidity/testing/snapshots/cst_output/UsingDirective/destructure_empty/input.sol:1]: Sure

File: crates/solidity/inputs/language/src/definition.rs (hunk 1 above)
Comment (Xanewok) [reply crates/solidity/inputs/language/src/definition.rs]: Thanks, updated!

File: crates/solidity/inputs/language/src/definition.rs (hunk 2 above)
Comment (Xanewok) [reply crates/solidity/inputs/language/src/definition.rs]: Makes sense! Pushed an update

File: crates/solidity/testing/snapshots/cst_output/TupleDeconstructionStatement/empty/generated/0.4.11-success.yml (hunk 3 above)
Comment (OmarTawfik) [reply crates/solidity/testing/snapshots/cst_output/TupleDeconstructionStatement/empty/generated/0.4.11-success.yml:5]: I see. I suggest adding the issue # to this TODO comment here, if the bug already exists on main.

Comment (Xanewok) [reply crates/solidity/testing/snapshots/cst_output/TupleDeconstructionStatement/empty/generated/0.4.11-success.yml]: Added in https://github.com/NomicFoundation/slang/pull/667/commits/23f88047d61fb5b6b4a459f643029e4069e83aba

**Review by OmarTawfik — APPROVED** [review]

**Pull request merged**